src/functions.jl

```diff
@@ -0,0 +1,164 @@
+"""
+    map(f, ds, [key]; dims) -> KeyedDataset
+
+Apply function `f` to each component of the [`KeyedDataset`](@ref).
+Returns a new dataset with the same constraints, but new components.
+The function can be applied to a subselection of components via a [`Pattern`](@ref) `key` and/or `dims`.
+
+# Example
+```jldoctest
+julia> using AxisKeys, Statistics; using AxisSets: KeyedDataset;
+
+julia> ds = KeyedDataset(
+           :val1 => KeyedArray(zeros(3, 2); time=1:3, obj=[:a, :b]),
+           :val2 => KeyedArray(ones(3, 2); time=1:3, loc=[:x, :y]),
+       );
+
+julia> mean.((ds.val1, ds.val2))
+(0.0, 1.0)
+
+julia> r = map(a -> a .+ 1.0, ds; dims=:loc);
+
+julia> mean.((r.val1, r.val2))
+(0.0, 2.0)
+```
+"""
+function Base.map(f::Function, ds::KeyedDataset, key::Tuple; kwargs...)
+    return map(f, ds, Pattern(key); kwargs...)
+end
+
+function Base.map(f::Function, ds::KeyedDataset, key::Pattern=Pattern(:__); dims=:)
+    # Select any components where the key and dims match
+    # NOTE: If this is a common operation maybe we should working into the lookup syntax?
+    selected = filter(collect(keys(ds.data))) do k
+        names = dimnames(ds.data[k])
+        dimsmatch = (
+            dims === Colon() ||
+            (isa(dims, Symbol) && dims in names) ||
+            (!isa(dims, Symbol) && dims ⊆ names)
+        )
+        return dimsmatch && k in key
+    end
+
+    return map!(f, deepcopy(ds), ds(k -> k in selected))
+end
+
+function Base.map!(f::Function, dest::KeyedDataset, src::KeyedDataset)
+    # Mutate the underlying data field to avoid calling `validate`
+    # on each insertion
+    data = dest.data
+    for (k, v) in src.data
+        data[k] = f(v)
+    end
+
+    validate(dest)
+    return dest
+end
+
+"""
+    mapslices(f, ds, [key]; dims) -> KeyedDataset
+
+Apply the `mapslices` call to each of the desired components and returns a new [`KeyedDataset`](@ref).
+The desired components can be selected via a [`Pattern`](@ref) `key` and/or `dims`.
+
+# Example
+```jldoctest
+julia> using AxisKeys, Statistics; using AxisSets: KeyedDataset;
+
+julia> ds = KeyedDataset(
+           :val1 => KeyedArray(zeros(3, 2); time=1:3, obj=[:a, :b]),
+           :val2 => KeyedArray(ones(3, 2); time=1:3, loc=[:x, :y]),
+       );
+
+julia> r = mapslices(sum, ds; dims=:time);
+
+julia> mean.((r.val1, r.val2))
```

Comment:
How does this relate to the :time dim? I think this example could be a bit clearer

Reply:
> How does this relate to the :time dim?

Output is 3 because we've applied `sum` along the time dimension which is length 3.

> I think this example could be a bit clearer

Ideally, I could just show the resulting `KeyedDataset`, but doctests are annoyingly finicky about white space with the printed `KeyedArrays`. I can try to get individual components to display consistently, but it'll likely be brittle.

Reply:
Yeah, I think I'm gonna go back to dropping any examples with `KeyedArray` and `KeyedDataset` printing. `jldoctest` is too much of a pain to bother trying to make this work consistently on the REPL, in tests and when building documentation. If folks want clearer examples they can follow the `@repl` output in the demo docs or read the tests.

Reply:
Opted to drop the `dims` example in favour of a pattern. I'm now showing the parent arrays to avoid `jldoctest` issues.